File: scripts/backend-python/subscription_service.py
"""Native Create Subscription - POST /subscription/create.

Doc: https://www.paytmpayments.com/docs/api/initiate-subscription-api
Defaults baked in:
- subscriptionPaymentMode: "UNKNOWN"  (Paytm renders all enabled rails)
- txnAmount.value: "2.00"             (CC/DC mandates require > Rs.1)
- subscriptionGraceDays: "3"          (CC/DC max)
- subscriptionStartDate: today (IST)
- subscriptionEnableRetry: "0"        (retry off; subscriptionRetryCount omitted)
"""
import json
import re
import secrets
from datetime import datetime, timedelta, timezone
from decimal import Decimal, ROUND_HALF_UP
from typing import Any, Optional

import requests
from paytmchecksum import PaytmChecksum

from paytm_config import get_paytm_config
from paytm_service import PaytmError

IST = timezone(timedelta(hours=5, minutes=30))


def _sanitize_cust_id(s: Optional[str]) -> str:
    return re.sub(r"[^a-zA-Z0-9_]", "_", s or "CUST_DEMO")


def _today_ist() -> str:
    return datetime.now(IST).strftime("%Y-%m-%d")


def _plus_one_year(yyyymmdd: str) -> str:
    d = datetime.strptime(yyyymmdd, "%Y-%m-%d")
    return d.replace(year=d.year + 1).strftime("%Y-%m-%d")


_TWO_PLACES = Decimal("0.01")


def _normalize_amount(amount: Any, min_rupees: float = 2.0) -> str:
    """Two-decimal currency normalization using Decimal (avoids binary-float drift)."""
    floor = Decimal(str(min_rupees)).quantize(_TWO_PLACES)
    raw = ("" if amount is None else str(amount)).strip()
    try:
        d = Decimal(raw) if raw else floor
    except Exception:
        d = floor
    if d < floor:
        d = floor
    return str(d.quantize(_TWO_PLACES, rounding=ROUND_HALF_UP))


def create_subscription(
    *,
    amount: Any = "2.00",
    renewal_amount: Optional[Any] = None,
    cust_id: Optional[str] = None,
    mobile: Optional[str] = None,
    email: Optional[str] = None,
    first_name: Optional[str] = None,
    last_name: Optional[str] = None,
    frequency: str = "1",
    frequency_unit: str = "MONTH",
    amount_type: str = "FIX",       # FIX | VARIABLE
    max_amount: Optional[Any] = None,
    start_date: Optional[str] = None,
    expiry_date: Optional[str] = None,
    grace_days: str = "3",
    payment_mode: str = "UNKNOWN",  # CC | DC | BANK_MANDATE | UNKNOWN
    mandate_type: Optional[str] = None,  # E_MANDATE | PAPER_MANDATE - only with BANK_MANDATE
    order_id: Optional[str] = None,
    server_base_url: str = "",
) -> dict:
    cfg = get_paytm_config()
    if not cfg["mid"]:
        raise PaytmError("MISSING_MID", "Missing PAYTM_MID")
    if not cfg["merchant_key"]:
        raise PaytmError("MISSING_MERCHANT_KEY", "Missing PAYTM_MERCHANT_KEY")

    order_id = (order_id or "").strip() or ("SUB_" + secrets.token_hex(10).upper())
    trace_id = "TRC_" + secrets.token_hex(10).upper()
    start = (start_date or "").strip() or _today_ist()
    expiry = (expiry_date or "").strip() or _plus_one_year(start)
    callback_url = cfg["callback_url"] or f"{server_base_url.rstrip('/')}/paytm/callback"

    user_info: dict = {"custId": _sanitize_cust_id(cust_id)}
    if mobile and mobile.strip(): user_info["mobile"] = mobile.strip()
    if email and email.strip(): user_info["email"] = email.strip()
    if first_name and first_name.strip(): user_info["firstName"] = first_name.strip()
    if last_name and last_name.strip(): user_info["lastName"] = last_name.strip()

    body: dict = {
        "requestType": "NATIVE_SUBSCRIPTION",
        "mid": cfg["mid"],
        "orderId": order_id,
        "websiteName": cfg["website_name"],
        "txnAmount": {"value": _normalize_amount(amount), "currency": "INR"},
        "subscriptionPaymentMode": payment_mode,
        "subscriptionAmountType": amount_type,
        "subscriptionFrequency": str(frequency),
        "subscriptionFrequencyUnit": frequency_unit,
        "subscriptionStartDate": start,
        "subscriptionExpiryDate": expiry,
        "subscriptionGraceDays": str(grace_days),
        "subscriptionEnableRetry": "0",
        "userInfo": user_info,
        "callbackUrl": callback_url,
    }

    if amount_type == "VARIABLE":
        if not max_amount:
            raise PaytmError("MISSING_MAX_AMOUNT",
                             "subscriptionMaxAmount required for VARIABLE amount type")
        body["subscriptionMaxAmount"] = _normalize_amount(max_amount)
    elif max_amount:
        body["subscriptionMaxAmount"] = _normalize_amount(max_amount)
    if renewal_amount:
        body["renewalAmount"] = _normalize_amount(renewal_amount)
    if payment_mode == "BANK_MANDATE":
        body["mandateType"] = mandate_type or "E_MANDATE"

    signature = PaytmChecksum.generateSignature(json.dumps(body), cfg["merchant_key"])
    head = {
        "clientId": cfg["client_id"],
        "channelId": "WEB",
        "version": "v1",
        "requestTimestamp": str(int(datetime.now(timezone.utc).timestamp() * 1000)),
        "signature": signature,
    }

    url = f"{cfg['subscription_create_url']}?mid={cfg['mid']}&orderId={order_id}&traceId={trace_id}"
    r = requests.post(url, json={"head": head, "body": body},
                      headers={"Content-Type": "application/json"}, timeout=15)
    text = r.text
    if not r.ok:
        raise _upstream("SUBSCRIPTION_HTTP_ERROR",
                        f"subscription/create HTTP {r.status_code}", order_id, text)

    data = json.loads(text)
    info = (data.get("body") or {}).get("resultInfo") or {}
    if info.get("resultStatus") != "S":
        raise _upstream("SUBSCRIPTION_FAILED",
                        info.get("resultMsg") or "subscription/create failed", order_id, text)

    return {
        "orderId": order_id,
        "traceId": trace_id,
        "txnToken": (data.get("body") or {}).get("txnToken"),
        "subscriptionId": (data.get("body") or {}).get("subscriptionId"),
        "amount": body["txnAmount"]["value"],
        "mid": cfg["mid"],
        "tokenType": "TXN_TOKEN",
    }


def _upstream(code: str, message: str, order_id: str, raw: str) -> PaytmError:
    paytm: dict = {}
    try:
        info = ((json.loads(raw) or {}).get("body") or {}).get("resultInfo") or {}
        for k in ("resultStatus", "resultCode", "resultMsg"):
            if info.get(k):
                paytm[k] = info[k]
    except Exception:
        pass
    return PaytmError(code, message, http_status=502, order_id=order_id, paytm=paytm or None)


# ---------------------------------------------------------------------------
# Post-consent recurring lifecycle.
#
# IMPORTANT host split: subscription/create is the ONLY endpoint on the
# /theia/api/v1 prefix (and only on production). EVERY management API below
# lives on the NON-/theia/ host on BOTH environments - see paytm_config.py.
# Reusing subscription_create_url for these returns 404 / HTML on prod.
#
# The `head` envelope differs per endpoint; each function sets only the fields
# that endpoint requires (create's head does NOT carry over).
# ---------------------------------------------------------------------------


def _post_signed(url: str, body: dict, head_extra: dict, cfg: dict) -> dict:
    """Sign `body` and POST {head, body} to `url`. `head_extra` carries the
    per-endpoint head fields (e.g. {"tokenType": "AES"})."""
    signature = PaytmChecksum.generateSignature(json.dumps(body), cfg["merchant_key"])
    head = {**head_extra, "signature": signature}
    r = requests.post(url, json={"head": head, "body": body},
                      headers={"Content-Type": "application/json"}, timeout=15)
    text = r.text
    if not r.ok:
        raise _upstream("SUBSCRIPTION_HTTP_ERROR", f"{url} HTTP {r.status_code}",
                        body.get("orderId", ""), text)
    try:
        return json.loads(text)
    except Exception:
        # A non-JSON (HTML) body almost always means the wrong host/prefix was used.
        raise _upstream("SUBSCRIPTION_NON_JSON",
                        f"{url} returned non-JSON (wrong host/prefix?)",
                        body.get("orderId", ""), text)


def check_subscription_status(*, subs_id: Optional[str] = None,
                              order_id: Optional[str] = None,
                              cust_id: Optional[str] = None) -> dict:
    """POST /subscription/checkStatus - the only reliable cancel/expiry signal.
    head: tokenType "AES" + signature. A cancelled mandate is status:"REJECT" +
    subStatus MERCHANT_CANCELLED/USER_CANCELLED - read subStatus to interpret."""
    cfg = get_paytm_config()
    body: dict = {"mid": cfg["mid"]}
    if subs_id:
        body["subsId"] = subs_id
    if order_id:
        body["orderId"] = order_id
    if cust_id:
        body["custId"] = _sanitize_cust_id(cust_id)
    data = _post_signed(cfg["subscription_check_status_url"], body, {"tokenType": "AES"}, cfg)
    b = data.get("body") or {}
    sub_status = b.get("subStatus") or ""
    return {
        "subsId": b.get("subsId"),
        "status": b.get("status"),
        "subStatus": b.get("subStatus"),
        "lastOrderStatus": b.get("lastOrderStatus"),
        "expiryDate": b.get("expiryDate"),
        "cancelled": b.get("status") == "REJECT" and sub_status.endswith("CANCELLED"),
        "active": b.get("status") == "ACTIVE" and sub_status == "ACTIVE",
        "raw": b,
    }


def renew_subscription(*, subscription_id: str, amount: Any,
                       order_id: Optional[str] = None) -> dict:
    """POST /subscription/renew - trigger one recurring debit. head: bare signature.
    Use a fresh order_id per debit, then confirm with get_order_status(order_id)."""
    cfg = get_paytm_config()
    order_id = (order_id or "").strip() or ("RENEW_" + secrets.token_hex(8).upper())
    body = {
        "mid": cfg["mid"],
        "subscriptionId": subscription_id,
        "orderId": order_id,
        "txnAmount": {"value": _normalize_amount(amount), "currency": "INR"},
    }
    data = _post_signed(cfg["subscription_renew_url"], body, {}, cfg)
    return {"orderId": order_id, "raw": data.get("body") or data}


def pre_notify_subscription(*, subscription_id: str, amount: Any,
                            order_id: Optional[str] = None,
                            scheduled_execution_date: Optional[str] = None) -> dict:
    """POST /subscription/preNotify - NPCI pre-debit notice. You MUST call this
    before every debit; Paytm does not auto-handle it. head: tokenType "AES" + signature."""
    cfg = get_paytm_config()
    order_id = (order_id or "").strip() or ("PRENOTIFY_" + secrets.token_hex(8).upper())
    body: dict = {
        "mid": cfg["mid"],
        "subscriptionId": subscription_id,
        "orderId": order_id,
        "txnAmount": {"value": _normalize_amount(amount), "currency": "INR"},
    }
    if scheduled_execution_date:
        body["subscriptionScheduledExecutionDate"] = scheduled_execution_date
    data = _post_signed(cfg["subscription_pre_notify_url"], body, {"tokenType": "AES"}, cfg)
    return {"orderId": order_id, "raw": data.get("body") or data}


def pre_notify_status(*, subscription_id: str, order_id: str) -> dict:
    """POST /subscription/preNotify/status.
    head: clientId + tokenType "AES" + version + signature."""
    cfg = get_paytm_config()
    body = {"mid": cfg["mid"], "subscriptionId": subscription_id, "orderId": order_id}
    data = _post_signed(
        cfg["subscription_pre_notify_status_url"], body,
        {"clientId": cfg["client_id"], "version": "v1", "tokenType": "AES"}, cfg,
    )
    return data.get("body") or data


def cancel_subscription(*, subscription_id: str) -> dict:
    """POST /subscription/cancel - terminal. head: signature + tokenType "AES"."""
    cfg = get_paytm_config()
    body = {"mid": cfg["mid"], "subscriptionId": subscription_id}
    data = _post_signed(cfg["subscription_cancel_url"], body, {"tokenType": "AES"}, cfg)
    return data.get("body") or data


def get_order_status(*, order_id: str) -> dict:
    """POST /v3/order/status - per-charge transaction status (NOT mandate state).
    head: bare signature. Same host on both envs (non-/theia/, non-versioned)."""
    cfg = get_paytm_config()
    body = {"mid": cfg["mid"], "orderId": order_id}
    data = _post_signed(cfg["order_status_url"], body, {}, cfg)
    return data.get("body") or data

File: scripts/backend-node/subscriptionService.js
// Native Create Subscription - POST /subscription/create
// Doc: https://www.paytmpayments.com/docs/api/initiate-subscription-api
// Defaults baked in:
//   - subscriptionPaymentMode: "UNKNOWN"  (Paytm renders all enabled rails)
//   - txnAmount.value: "2.00"             (CC/DC mandates require > ₹1)
//   - subscriptionGraceDays: "3"          (CC/DC max)
//   - subscriptionStartDate: today (IST)
//   - subscriptionEnableRetry: "0"        (retry off; subscriptionRetryCount omitted)
import crypto from "node:crypto";
import PaytmChecksum from "paytmchecksum";
import { getPaytmConfig } from "./paytmConfig.js";

function sanitizeCustId(s) {
  return (s || "CUST_DEMO").replace(/[^a-zA-Z0-9_]/g, "_");
}

function todayIST() {
  // YYYY-MM-DD in IST regardless of server timezone.
  const offsetMs = 5.5 * 60 * 60 * 1000;
  return new Date(Date.now() + offsetMs).toISOString().slice(0, 10);
}

function todayISTddmmyyyy() {
  // DD-MM-YYYY in IST — preNotify.txnDate format (differs from the YYYY-MM-DD dates).
  const [y, m, d] = todayIST().split("-");
  return `${d}-${m}-${y}`;
}

function plusOneYear(yyyymmdd) {
  const [y, m, d] = yyyymmdd.split("-").map(Number);
  return `${y + 1}-${String(m).padStart(2, "0")}-${String(d).padStart(2, "0")}`;
}

function normalizeAmount(amount, minRupees = 2) {
  // Two-decimal currency normalization, integer-paise rounding to avoid
  // binary-float drift. Returns a string (Paytm subscription `txnAmount.value`
  // contract).
  const n = Number(String(amount ?? "").trim());
  if (!Number.isFinite(n) || n < minRupees) return minRupees.toFixed(2);
  const rounded = Math.round((n + Number.EPSILON) * 100) / 100;
  return rounded.toFixed(2);
}

export async function createSubscription({
  amount,                           // first-debit amount (string/number, ≥ ₹2 for cross-rail)
  renewalAmount,                    // optional; only set if you want a different recurring amount
  custId,
  mobile,
  email,
  firstName,
  lastName,
  frequency = "1",                  // number of units per cycle
  frequencyUnit = "MONTH",          // DAY | WEEK | MONTH | YEAR | ONDEMAND
  amountType = "FIX",               // FIX | VARIABLE (VARIABLE requires subscriptionMaxAmount)
  maxAmount,                        // required when amountType === "VARIABLE"
  startDate,                        // YYYY-MM-DD; defaults to today
  expiryDate,                       // YYYY-MM-DD; defaults to startDate + 1 year
  graceDays = "3",
  paymentMode = "UPI",              // UPI (default) | CC | DC | BANK_MANDATE | UNKNOWN
                                    // "UNKNOWN" can render an empty checkout on some prod MIDs.
                                    // Note: UPI accepts only WEEK/MONTH/YEAR frequency units (not DAY).
  mandateType,                      // E_MANDATE | PAPER_MANDATE - only with BANK_MANDATE
  orderId: callerOrderId,
  serverBaseUrl,
}) {
  const cfg = getPaytmConfig();
  if (!cfg.mid) throw asError("MISSING_MID", "Missing PAYTM_MID");
  if (!cfg.merchantKey) throw asError("MISSING_MERCHANT_KEY", "Missing PAYTM_MERCHANT_KEY");

  const orderId = callerOrderId?.trim()
    || `SUB_${crypto.randomBytes(10).toString("hex").toUpperCase()}`;
  const traceId = `TRC_${crypto.randomBytes(10).toString("hex").toUpperCase()}`;
  const start = startDate?.trim() || todayIST();
  const expiry = expiryDate?.trim() || plusOneYear(start);
  const callbackUrl = cfg.callbackUrl?.trim()
    || `${(serverBaseUrl || cfg.callbackBase).replace(/\/+$/, "")}/paytm/callback`;

  const body = {
    requestType: "NATIVE_SUBSCRIPTION",
    mid: cfg.mid,
    orderId,
    websiteName: cfg.websiteName,
    txnAmount: { value: normalizeAmount(amount), currency: "INR" },
    subscriptionPaymentMode: paymentMode,
    subscriptionAmountType: amountType,
    subscriptionFrequency: String(frequency),
    subscriptionFrequencyUnit: frequencyUnit,
    subscriptionStartDate: start,
    subscriptionExpiryDate: expiry,
    subscriptionGraceDays: String(graceDays),
    subscriptionEnableRetry: "0",
    userInfo: {
      custId: sanitizeCustId(custId),
      ...(mobile?.trim() ? { mobile: mobile.trim() } : {}),
      ...(email?.trim() ? { email: email.trim() } : {}),
      ...(firstName?.trim() ? { firstName: firstName.trim() } : {}),
      ...(lastName?.trim() ? { lastName: lastName.trim() } : {}),
    },
    callbackUrl,
  };
  if (amountType === "VARIABLE") {
    if (!maxAmount) throw asError("MISSING_MAX_AMOUNT", "subscriptionMaxAmount required for VARIABLE amount type");
    body.subscriptionMaxAmount = normalizeAmount(maxAmount);
  } else if (maxAmount) {
    body.subscriptionMaxAmount = normalizeAmount(maxAmount);
  }
  if (renewalAmount) body.renewalAmount = normalizeAmount(renewalAmount);
  if (paymentMode === "BANK_MANDATE") {
    body.mandateType = mandateType || "E_MANDATE";
  }

  const signature = await PaytmChecksum.generateSignature(JSON.stringify(body), cfg.merchantKey);
  const head = {
    clientId: cfg.clientId,
    channelId: "WEB",
    version: "v1",
    requestTimestamp: String(Date.now()),
    signature,
  };

  const url = new URL(cfg.subscriptionCreateUrl);
  url.searchParams.set("mid", cfg.mid);
  url.searchParams.set("orderId", orderId);
  url.searchParams.set("traceId", traceId);

  const r = await fetch(url.toString(), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ head, body }),
  });
  const text = await r.text();
  if (!r.ok) throw upstream("SUBSCRIPTION_HTTP_ERROR", `subscription/create HTTP ${r.status}`, orderId, text);

  const json = JSON.parse(text);
  const status = json?.body?.resultInfo?.resultStatus;
  if (status !== "S") {
    throw upstream("SUBSCRIPTION_FAILED", json?.body?.resultInfo?.resultMsg || "subscription/create failed", orderId, text);
  }
  return {
    orderId,
    traceId,
    txnToken: json?.body?.txnToken,
    subscriptionId: json?.body?.subscriptionId,
    amount: body.txnAmount.value,
    mid: cfg.mid,
    tokenType: "TXN_TOKEN",
  };
}

function asError(code, message) {
  const e = new Error(message);
  e.code = code;
  return e;
}

function upstream(code, message, orderId, raw) {
  const e = new Error(message);
  e.code = code;
  e.httpStatus = 502;
  e.orderId = orderId;
  try {
    const info = JSON.parse(raw)?.body?.resultInfo;
    if (info) e.paytm = { resultStatus: info.resultStatus, resultCode: info.resultCode, resultMsg: info.resultMsg };
  } catch {}
  return e;
}

// ---------------------------------------------------------------------------
// Post-consent recurring lifecycle.
//
// IMPORTANT host split: subscription/create is the ONLY endpoint on the
// /theia/api/v1 prefix (and only on production). EVERY management API below
// lives on the NON-/theia/ host on BOTH environments — see paytmConfig.js.
// Reusing cfg.subscriptionCreateUrl for these returns 404 / HTML on prod.
//
// The `head` envelope differs per endpoint; each function below sets the
// fields that endpoint actually requires (create's head does NOT carry over).
// ---------------------------------------------------------------------------

// Signs `body` and POSTs `{ head, body }` to `url`. `headExtra` carries the
// per-endpoint head fields (e.g. { tokenType: "AES" }).
async function postSigned(url, body, headExtra, cfg) {
  const signature = await PaytmChecksum.generateSignature(JSON.stringify(body), cfg.merchantKey);
  const head = { ...headExtra, signature };
  const r = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ head, body }),
  });
  const text = await r.text();
  if (!r.ok) throw upstream("SUBSCRIPTION_HTTP_ERROR", `${url} HTTP ${r.status}`, body.orderId, text);
  try {
    return JSON.parse(text);
  } catch {
    // A non-JSON (HTML) body almost always means the wrong host/prefix was used.
    throw upstream("SUBSCRIPTION_NON_JSON", `${url} returned non-JSON (wrong host/prefix?)`, body.orderId, text);
  }
}

// POST /subscription/checkStatus — the only reliable cancel/expiry signal.
// head: tokenType "AES" + signature. Identify by subsId (or orderId+custId).
// Returns { status, subStatus, ... } — read subStatus to interpret a REJECT.
export async function checkSubscriptionStatus({ subsId, orderId, custId } = {}) {
  const cfg = getPaytmConfig();
  const body = { mid: cfg.mid };
  if (subsId) body.subsId = subsId;
  if (orderId) body.orderId = orderId;
  if (custId) body.custId = sanitizeCustId(custId);
  const json = await postSigned(cfg.subscriptionCheckStatusUrl, body, { tokenType: "AES" }, cfg);
  const b = json?.body || {};
  return {
    subsId: b.subsId,
    status: b.status,
    subStatus: b.subStatus,
    lastOrderStatus: b.lastOrderStatus,
    expiryDate: b.expiryDate,
    // A cancelled mandate is status:"REJECT" + subStatus:MERCHANT_CANCELLED/USER_CANCELLED.
    cancelled: b.status === "REJECT" && /CANCELLED$/.test(b.subStatus || ""),
    active: b.status === "ACTIVE" && b.subStatus === "ACTIVE",
    raw: b,
  };
}

// POST /subscription/renew — trigger one recurring debit. head: bare signature.
// mid + orderId MUST also be in the query string (else 1007 / 2014). A successful
// preNotify must precede renew, else renew returns 3054.
// Use a fresh orderId per debit, then confirm with getOrderStatus(orderId).
export async function renewSubscription({ subscriptionId, amount, orderId: callerOrderId }) {
  const cfg = getPaytmConfig();
  const orderId = callerOrderId?.trim()
    || `RENEW_${crypto.randomBytes(8).toString("hex").toUpperCase()}`;
  const body = {
    mid: cfg.mid,
    subscriptionId,
    orderId,
    txnAmount: { value: normalizeAmount(amount), currency: "INR" },
  };
  const url = new URL(cfg.subscriptionRenewUrl);
  url.searchParams.set("mid", cfg.mid);
  url.searchParams.set("orderId", orderId);
  const json = await postSigned(url.toString(), body, {}, cfg);
  return { orderId, raw: json?.body || json };
}

// POST /subscription/preNotify — NPCI pre-debit notice. You MUST call this
// before every debit; Paytm does not auto-handle it. head: tokenType "AES" + signature.
// Field quirks (live prod): txnAmount is a FLAT STRING (not {value,currency});
// txnDate = today in DD-MM-YYYY; subsId + referenceId = subscriptionId; txnMessage /
// merchantName / merchantLogoUrl required. Call within 1–7 days before the billing date.
export async function preNotifySubscription({
  subscriptionId, amount, orderId: callerOrderId,
  scheduledExecutionDate,           // billing date, YYYY-MM-DD
  txnMessage = "Subscription renewal",
  merchantName = "Subscription",
  merchantLogoUrl = "",
}) {
  const cfg = getPaytmConfig();
  const orderId = callerOrderId?.trim()
    || `PRENOTIFY_${crypto.randomBytes(8).toString("hex").toUpperCase()}`;
  const body = {
    mid: cfg.mid,
    subscriptionId,
    subsId: subscriptionId,
    referenceId: subscriptionId,
    orderId,
    txnAmount: normalizeAmount(amount),   // flat string, NOT an object
    txnDate: todayISTddmmyyyy(),          // today, DD-MM-YYYY
    txnMessage,
    merchantName,
    merchantLogoUrl,
  };
  if (scheduledExecutionDate) body.subscriptionScheduledExecutionDate = scheduledExecutionDate;
  const json = await postSigned(cfg.subscriptionPreNotifyUrl, body, { tokenType: "AES" }, cfg);
  return { orderId, raw: json?.body || json };
}

// POST /subscription/preNotify/status. head: clientId + tokenType "AES" + version + signature.
export async function preNotifyStatus({ subscriptionId, orderId }) {
  const cfg = getPaytmConfig();
  const body = { mid: cfg.mid, subscriptionId, orderId };
  const json = await postSigned(
    cfg.subscriptionPreNotifyStatusUrl,
    body,
    { clientId: cfg.clientId, version: "v1", tokenType: "AES" },
    cfg,
  );
  return json?.body || json;
}

// POST /subscription/cancel — terminal. head: signature + tokenType "AES".
// subscriptionId must also be in the query string; subsId (same value) required in body
// (else 400 "Subscription Id field can not be empty").
export async function cancelSubscription({ subscriptionId }) {
  const cfg = getPaytmConfig();
  const body = { mid: cfg.mid, subscriptionId, subsId: subscriptionId };
  const url = new URL(cfg.subscriptionCancelUrl);
  url.searchParams.set("mid", cfg.mid);
  url.searchParams.set("subscriptionId", subscriptionId);
  const json = await postSigned(url.toString(), body, { tokenType: "AES" }, cfg);
  return json?.body || json;
}

// POST /v3/order/status — per-charge transaction status (NOT mandate state).
// head: bare signature. Same host on both envs (non-/theia/, non-versioned).
export async function getOrderStatus({ orderId }) {
  const cfg = getPaytmConfig();
  const body = { mid: cfg.mid, orderId };
  const json = await postSigned(cfg.orderStatusUrl, body, {}, cfg);
  return json?.body || json;
}
